### ISSUE TITLE
A method to set a new CROW_STATIC_DIRECTORY at runtime

### ISSUE THREAD

**Issue by joelguittet** [CrowCpp/Crow#414]
Hello,

I would like to suggest an enhancement about CROW_STATIC_DIRECTORY because this definition is not always very convenient for building.

CROW_STATIC_DIRECTORY should be the default value for example and I would like to have a method to set a new value at runtime.

As done for `crow::mustache::set_global_base` for example. If the idea is appreciated I can create a pull request.

Joel

**Comment by The-EDev** [comment]
Hi, While I do agree migrating Crow away from macros is a good idea (especially when it comes to being able to run more than one app per process), I'm not sure allowing runtime changes of the static directory is a good idea. In fact it was never the intention for `mustache::set_global_base` to be used during runtime, only `mustache::set_base`.

I'm defining "runtime" here as being called after the execution of `crow::run()`.

**Comment by joelguittet** [comment]
Hello @The-EDev 
Agree with you after run() is called allowing changing the directory is probably not a good idea.
Removing this kind of macros can benefit people who want to build a library (I mean a .so file for example), which is much more  convenient usually (Im' working on embedded system with Yocto and integrating crow in the build to populate the sdk should be a fantastic idea :-) ).
Joel

**Comment by The-EDev** [comment]
Looking into this issue, `CROW_STATIC_DIRECTORY` was easily convertible to a function. the main problem was in the fact that `CROW_STATIC_ENDPOINT` is of type `black_magic::const_str` Which absolutely does not support runtime configuration. I'm not sure what trick I can employ to allow this, but if we are to take the language at face value, `CROW_STATIC_ENDPOINT` must stay a macro. Unless some C++11 version of a macro is possible.

**Comment by The-EDev** [comment]
Since this issue only references `CROW_STATIC_DIRECTORY`, the fix should be easy.. Do you think `CROW_STATIC_ENDPOINT` should also be part of this issue @joelguittet?

**Comment by joelguittet** [comment]
Agree, probably a common strategy and solution should be applied to several parameters.

**Comment by xdanieldev** [comment]
My solution is:

// Fi rst disable default crow add_static_dir
#define CROW_DISABLE_STATIC_DIR

/**
 * \brief Adds a static route to serve files from a directory
 *
 * \details Sets up a route that serves static files from a specified directory.
 * The endpoint MUST follow the pattern "/something/<path>" where <path> is the 
 * file path parameter (e.g., "/static/<path>" or "/assets/<path>").
 *
 * \tparam Middlewares Middleware types used in the Crow application
 * 
 * \param app Reference to the Crow application
 * \param static_endpoint Must be in the format "/something/<path>" (e.g., "/static/<path>")
 * \param static_dir Local directory path that contains the static files
 *
 * \example
 * crow::SimpleApp app;
 * add_static_dir(app, "/static/<path>", "./public");  // Valid
 * add_static_dir(app, "/files/images/<path>", "./images");  // Valid
 * // add_static_dir(app, "/static", "./public");  // Invalid - missing <path>
 */
template<typename... Middlewares>
void add_static_dir(crow::Crow<Middlewares...>& app,
                     std::string static_endpoint,
                     std::string static_dir) {
    constexpr uint64_t tag = 5; // <path> enum
    
    app.template route<tag>(std::move(static_endpoint))
    ([dir = std::move(static_dir)](crow::response& res, std::string file_path_partial) {
        crow::utility::sanitize_filename(file_path_partial);
        res.set_static_file_info_unsafe(dir + file_path_partial);
        res.end();
    });
}

Example code:
crow::SimpleApp app;
add_static_dir(app, "/public/<path>", "/home/user/Public/");


Caveat:

define CROW_DISABLE_STATIC_DIR disable also "add_blueprint":

crow/app.h:
  #ifndef CROW_DISABLE_STATIC_DIR
              add_blueprint();
              add_static_dir();
  #endif
